src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1275,39 +1279,47 @@ private ArrayTypeSymbol VisitArrayInitializer(BoundArrayCreation node)
                 resultBuilder.Add(VisitRvalueWithResult(element));
             }
 
+            bool checkConversions;
             // Consider recording in the BoundArrayCreation
             // whether the array was implicitly typed, rather than relying on syntax.
             if (node.Syntax.Kind() == SyntaxKind.ImplicitArrayCreationExpression)
             {
-                var resultTypes = resultBuilder.ToImmutable();
-                // https://github.com/dotnet/roslyn/issues/27961 Initial binding calls InferBestType(ImmutableArray<BoundExpression>, ...)
-                // overload. Why are we calling InferBestType(ImmutableArray<TypeSymbolWithAnnotations>, ...) here?
-                // https://github.com/dotnet/roslyn/issues/27961 InferBestType(ImmutableArray<BoundExpression>, ...)
-                // uses a HashSet<TypeSymbol> to reduce the candidates to the unique types before comparing.
-                // Should do the same here.
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                // If there are error types, use the first error type. (Matches InferBestType(ImmutableArray<BoundExpression>, ...).)
-                var bestType = resultTypes.FirstOrDefault(t => !t.IsNull && t.IsErrorType());
-                if (bestType.IsNull)
+                if (node.HasErrors)
                 {
-                    bestType = BestTypeInferrer.InferBestType(resultTypes, _conversions, useSiteDiagnostics: ref useSiteDiagnostics);
+                    elementType = elementType.SetUnknownNullabilityForReferenceTypes();
+                    checkConversions = false;
                 }
                 else
                 {
-                    // Mark the error type as null-oblivious.
-                    bestType = TypeSymbolWithAnnotations.Create(bestType.TypeSymbol);
-                }
-                // https://github.com/dotnet/roslyn/issues/27961 Report a special ErrorCode.WRN_NoBestNullabilityArrayElements
-                // when InferBestType fails, and avoid reporting conversion warnings for each element in those cases.
-                // (See similar code for conditional expressions: ErrorCode.WRN_NoBestNullabilityConditionalExpression.)
-                if (!bestType.IsNull)
-                {
-                    elementType = bestType;
+                    var placeholderBuilder = ArrayBuilder<BoundExpression>.GetInstance(n);
+                    for (int i = 0; i < n; i++)
+                    {
+                        placeholderBuilder.Add(CreatePlaceholderIfNecessary(elementBuilder[i], resultBuilder[i]));
+                    }
+                    var placeholders = placeholderBuilder.ToImmutableAndFree();
+                    bool hadNullabilityMismatch;
+                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                    var bestType = BestTypeInferrer.InferBestType(placeholders, _conversions, out _, out hadNullabilityMismatch, ref useSiteDiagnostics);
```

Comment:
>useSiteDiagnostics [](start = 133, length = 18)

Do we need to report use site diagnostics?

Reply:
No, not after initial binding.

---
In reply to: [225614804](https://github.com/dotnet/roslyn/pull/29727#discussion_r225614804) [](ancestors = 225614804)